src/framework/MOM_diag_buffers.F90

```diff
@@ -0,0 +1,421 @@
+  integer :: slot
+
+  slot = this%find_buffer_slot(id)
+  this%ids(slot) = 0
+end subroutine mark_available
+
+!> Return the slot of the buffer corresponding to the diagnostic id
+pure function find_buffer_slot(this, id) result(slot)
+  class(diag_buffer_base), intent(in) :: this !< The diagnostic buffer
+  integer, intent(in) :: id !< The diagnostic id
+
+  integer, dimension(1) :: temp
+  integer :: slot !< The slot in the buffer corresponding to the diagnostic id
+
+  if(allocated(this%ids)) then
+    !NOTE: Alternatively could do slot = SUM(findloc(...))
+    temp = findloc(this%ids(:), id)
+    slot = temp(1)
+  else
+    slot = 0
+  endif
+
+end function find_buffer_slot
+
+!> Grow the ids array by one
+subroutine grow_ids(this)
+  class(diag_buffer_base), intent(inout) :: this !< This buffer
+
+  integer, allocatable, dimension(:) :: temp
+  integer :: n
+
+  n = this%length
+
+  allocate(temp(n+1))
+  if(n>0) temp(1:n) = this%ids(:)
+  call move_alloc(temp, this%ids)
+end subroutine grow_ids
+
+!> Check whether the id already has a slot reserved. If not, find a new empty slot and if
+!! need be, grow the buffer.
+impure function check_capacity_by_id(this, id) result(slot)
+  class(diag_buffer_base), intent(inout) :: this !< This 2d buffer
+  integer,                 intent(in)    :: id   !< The diagnostic id
+  integer :: slot
+
+  slot = this%find_buffer_slot(id)
+  if(slot==0) then
+    ! Check to see if there is an open slot
+    if(allocated(this%ids)) slot = this%find_buffer_slot(0)
+    ! If slot is still 0, then the buffer must grow
+    if (slot==0) then
+      call this%grow()
+      slot = this%length
+    endif
+    this%ids(slot) = id
+  endif
+end function check_capacity_by_id
+
+!> Set the horizontal extents of the buffer
+subroutine set_horizontal_extents(this, is, ie, js, je)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+  integer,               intent(in)    :: is !< The start slot of the array i-direction
+  integer,               intent(in)    :: ie !< The end slot of the array i-direction
+  integer,               intent(in)    :: js !< The start slot of the array j-direction
+  integer,               intent(in)    :: je !< The end slot of the array j-direction
+
+  this%is = is ; this%ie = ie ; this%js = js ; this%je = je
+end subroutine set_horizontal_extents
+
+!> Set the vertical extent of the buffer
+subroutine set_vertical_extent(this, ks, ke)
+  class(diag_buffer_3d), intent(inout) :: this !< The diagnostic buffer
+  integer,               intent(in)    :: ks !< The start slot of the array i-direction
+  integer,               intent(in)    :: ke !< The end slot of the array i-direction
+
+  this%ks = ks; this%ke = ke
+end subroutine set_vertical_extent
+
+!> Grow a buffer for 2D arrays
+subroutine grow_2d(this)
+  class(diag_buffer_2d), intent(inout) :: this !< This 2d buffer
+
+  integer :: n
+  real, allocatable, dimension(:,:,:) :: temp
```

Comment:
Please add comments describing the purpose and units (perhaps `[arbitrary]` or '[A ~> a]` of this real variable.

Reply:
Comments have been added to indicate that these temporary arrays and the primary `buffer` in the type are arbitrary. The values (and their accordant units) assigned to these buffers will depend on their actual use within the codebase. Whether these represent scaled arbitrary units, depends on how they are filled from within the codebase. As you point out, in the case of their use as part of `post_data`, this may be the case, but may not in general hold.